services/imap/mail_receiver.go

```diff
@@ -0,0 +1,219 @@
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/models/unit"
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/queue"
+	"code.gitea.io/gitea/modules/setting"
+	comment_service "code.gitea.io/gitea/services/comments"
+)
+
+// mail read query
+var mailReadQueue queue.Queue
+
+// NewContext start received mail read queue service
+func NewContext() {
+	if setting.MailRecieveService == nil || mailReadQueue != nil {
+		return
+	}
+
+	mailReadQueue = queue.CreateQueue("mail_receive", func(data ...queue.Data) {
+		for _, datum := range data {
+			mail := datum.(*Mail)
+			if err := mail.LoadHeader([]string{"From", "To", "In-Reply-To", "References"}); err != nil {
+				log.Error("fetch mail header failed: %v", err)
+				continue
+			}
+
+			if len(mail.Heads["To"]) == 0 ||
+				len(mail.Heads["From"]) == 0 {
+				continue
+			}
+
+			if mail.Heads["To"][0].Address != setting.MailRecieveService.ReceiveEmail {
+				continue
+			}
+
+			log.Trace("start read email from %v", mail.Heads["From"][0].String())
+			if err := handleReceiveEmail(mail); err != nil {
+				log.Error("handleReceiveEmail(): %v", err)
+				continue
+			}
+			log.Trace("finished read email from %v", mail.Heads["From"][0].String())
+		}
+	}, &Mail{})
+
+	go graceful.GetManager().RunWithShutdownFns(mailReadQueue.Run)
+}
+
+func handleReceiveEmail(m *Mail) error {
+	fromEmail, ok := m.Heads["From"]
+	if !ok || len(fromEmail) < 1 {
+		return nil
+	}
+	from := fromEmail[0].Address
+	doer, err := models.GetUserByEmail(from)
+	if err != nil {
+		if models.IsErrUserNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("models.GetUserByEmail(%v): %v", from, err)
+	}
+
+	checkLink := ""
+
+	// check `In-Reply-To`
+	if links, ok := m.Heads["In-Reply-To"]; ok && links != nil {
+		for _, link := range links {
+			if strings.Contains(link.Address, setting.Domain) {
+				checkLink = link.Address
+				break
+			}
+		}
+	}
+
+	if len(checkLink) == 0 {
+		// check `References`
+		if links, ok := m.Heads["References"]; ok && links != nil {
+			for _, link := range links {
+				if strings.Contains(link.Address, setting.Domain) {
+					checkLink = link.Address
+					break
+				}
+			}
+		}
+	}
+
+	if len(checkLink) == 0 {
+		_ = m.SetRead(true)
+		return nil
+	}
+
+	splitLink := strings.SplitN(checkLink, "@", 2)
+	if len(splitLink) != 2 || splitLink[1] != setting.Domain {
+		_ = m.SetRead(true)
+		return nil
+	}
+
+	splitLink = strings.SplitN(splitLink[0], "?", 2)
+	if len(splitLink) != 2 {
+		_ = m.SetRead(true)
+		return nil
+	}
+
+	checkKey := splitLink[1]
+
+	splitLink = strings.SplitN(splitLink[0], "#", 2)
+	if len(splitLink) == 0 {
+		_ = m.SetRead(true)
+		return nil
+	}
+
+	splitLink = strings.SplitN(splitLink[0], "/", 4)
+	if len(splitLink) != 4 {
+		_ = m.SetRead(true)
+		return nil
+	}
+
+	if len(splitLink) != 4 ||
+		(splitLink[2] != "pulls" && splitLink[2] != "issues") {
```

Comment:
What about Gitea in a sub path like: `https://mydomain/gitea/org/repo/...`

Reply:
In curent design,``message-id`` not include ``sub path``.